RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,304 @@
+* `var result1 = ProcessAttachmentSets([a1, a2, a3], ...);` when first 3 executions are done
+* `var result2 = ProcessAttachmentSets(result1.Concat([a4]), ...);` when 4th execution is done
+* `var finalResult = ProcessAttachmentSets(result2.Concat([a5]), ...);` when last test execution is done
+
+If `IsAssociative` is `False` then Test Platform will wait for all test executions to finish and call `ProcessAttachmentSets` only once:
+* `var finalResult = ProcessAttachmentSets([a1, a2, a3, a4, a5], ...);`
+
+By default `IsAssociative` should be `False`, unless processing can take longer time and it's beneficial to start the process as soon as possible.
+
+
+
+2. Introduce a new `FinalizeMultiTestRunAsync` method in [IVsTestConsoleWrapper](https://github.com/microsoft/vstest/blob/master/src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs) interface:
+
+```
+/// <summary>
+/// Provides back all attachments to TestPlatform for additional processing (for example merging)
+/// </summary>
+/// <param name="attachments">Collection of attachments</param>
+/// <param name="multiTestRunCompleted">Indicates that all test executions are done and all data is provided</param>
+/// <param name="collectMetrics">Enables metrics collection</param>
+/// <param name="multiTestRunFinalizationCompleteEventsHandler">EventHandler to receive session complete event</param>
+/// <param name="cancellationToken">Cancellation token</param>        
+Task FinalizeMultiTestRunAsync(IEnumerable<AttachmentSet> attachments, bool multiTestRunCompleted, bool collectMetrics, IMultiTestRunFinalizationEventsHandler eventsHandler, CancellationToken cancellationToken);
+```
+
+Method can be used to start a new Multi Test Run Finalization process, which is reprocessing all data collector attachments passed as first argument using all available processors.
+
+
+3. Introduce a new `IMultiTestRunFinalizationEventsHandler` interface:
+```
+namespace Microsoft.VisualStudio.TestPlatform.ObjectModel.Client
+{
+    /// <summary>
+    /// Interface contract for handling multi test run finalization complete events
+    /// </summary>
+    public interface IMultiTestRunFinalizationEventsHandler : ITestMessageEventHandler
+    {
+        /// <summary>
+        /// Dispatch MultiTestRunFinalizationComplete event to listeners.
+        /// </summary>
+        /// <param name="finalizationCompleteEventArgs">Finalization Complete event args.</param>
+        /// <param name="attachments">Last set of processed attachment sets.</param>
+        void HandleMultiTestRunFinalizationComplete(MultiTestRunFinalizationCompleteEventArgs finalizationCompleteEventArgs, IEnumerable<AttachmentSet> lastChunk);
+
+        /// <summary>
+        /// Dispatch FinalisedAttachments event to listeners.
+        /// </summary>
+        /// <param name="attachments">Finalised attachment sets.</param>
+        void HandleFinalisedAttachments(IEnumerable<AttachmentSet> attachments);
```

Comment:
nit: perhaps this should be called HandleFinalizedAttachmentsChunk? otherwise, the word finalized suggests that this could be the final set of attachments as well.

Reply:
Changed to:    

    /// <summary>
        /// Dispatch ProcessedAttachmentsChunk event to listeners.
        /// </summary>
        /// <param name="attachments">Processed attachment sets.</param>
        void HandleProcessedAttachmentsChunk(IEnumerable<AttachmentSet> attachments);